src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -225,9 +218,116 @@ private void pre20Upgrade() throws Exception {
+     * @param currentMetaData the current index state in memory.
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStateOnDataNode(ClusterChangedEvent event, MetaData currentMetaData, MetaStateService metaStateService, ESLogger logger) {
+        Map<String, GatewayMetaState.IndexMetaWriteInfo> indicesToWrite = new HashMap<>();
+        RoutingNode thisNode = event.state().getRoutingNodes().node(event.state().nodes().localNodeId());
+        if (thisNode == null) {
+            // this needs some other handling
+            return indicesToWrite.values();
+        }
+        // iterate over all shards allocated on this node in the new cluster state but only write if ...
+        for (MutableShardRouting shardRouting : thisNode) {
+            IndexMetaData indexMetaData = event.state().metaData().index(shardRouting.index());
+            IndexMetaData currentIndexMetaData = maybeLoadIndexState(currentMetaData, indexMetaData, metaStateService, logger);
+            String writeReason = null;
+            // ... state persistence was disabled or index was newly created
+            if (currentIndexMetaData == null) {
+                writeReason = "freshly created";
+                // ... new shard is allocated on node (we could optimize here and make sure only written once and not for each shard per index -> do later)
+            } else if (shardRouting.initializing()) {
+                writeReason = "newly allocated on node";
+                // ... version changed
+            } else if (indexMetaData.version() != currentIndexMetaData.version()) {
+                writeReason = "version changed from [" + currentIndexMetaData.version() + "] to [" + indexMetaData.version() + "]";
+            }
+            if (writeReason != null) {
+                indicesToWrite.put(shardRouting.index(),
+                        new GatewayMetaState.IndexMetaWriteInfo(indexMetaData, currentIndexMetaData,
+                                writeReason));
+            }
+        }
+        return indicesToWrite.values();
+    }
+
+    /**
+     * Loads the current meta state for each index in the new cluster state and checks if it has to be persisted.
+     * Each index state that is part of the new cluster state will be considered even if this node has no shard of the
+     * index allocated on it. This is only run for master nodes.
+     *
+     * @param event           the cluster state event from which we figure out what is new in each index and should potentially be written
+     * @param currentMetaData the current index state in memory.
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStatesOnMaster(ClusterChangedEvent event, MetaData currentMetaData, MetaStateService metaStateService, ESLogger logger) {
```

Comment:
since we only use the `event().state()` can we just pass the state to it instead?